sbc_compassion/tools/import_letter_functions.py

```diff
@@ -21,13 +21,117 @@
 ##########################################################################
+def testline(env, line_vals, csv_file, doc_name):
+    """ Test each line and verify the data given by the csv.
+    If a line does not correspond to the csv, this function erases it.
+    :param env: Odoo environment
+    :param dict line_vals: Dictionnary that will be used to create line
+    :param One2many('test.import.letter.line') csv_file: CSV used to check
+    :param str doc_name: Document name
+    :returns: test (True if equal to csv), error
+    :rtype: bool, str
+    """
+    # str that will be returned
+    error = ''
+    csv_data = []
+    # creates a list containing all the csv data
+    for csv_ in csv_file:
+        if os.path.splitext(csv_.name)[1] != '.csv':
+            raise Warning(_("File '{}' is not a csv file".format(csv_.name)))
+        csv_string = base64.b64decode(csv_.datas).splitlines()
+        csv_data += read_csv(csv_string)
+    # check the letters
+    values = {}
+    test = True
+    for csv_line in csv_data:
+        if csv_line['name'] == doc_name:
+            values = csv_line
+            break
+    if values == {}:
+        raise Warning(
+            _("File '{}' not found in the csv files".format(doc_name)))
+    # first test: template
+    template_id = env.ref(
+        'sbc_compassion.'+values['template_id'])
+    tpl_detected = env['sponsorship.correspondence.template'].search(
+        [('id', '=', line_vals['template_id'])])
+    temp_test = template_id == tpl_detected
+    test = test and temp_test
+    if not temp_test:
+        error = error + 'template,\n'
+
+    # second: qr code
+    # partner
+    partner_temp = env['res.partner'].search([
+        ('id', '=', line_vals['partner_id'])])
+    # child
+    child_temp = env['compassion.child'].search([
+        ('id', '=', line_vals['child_id'])])
+    pcodega = partner_temp.ref
+    ccode = child_temp.code
+    temp_test = True
+    if values['barcode'] == '' and not (pcodega and ccode):
+        temp_test = True
+    # xor (if both pcodega and ccode are False => barcode not detected)
+    elif bool(pcodega) ^ bool(ccode):
+        temp_test = True
+    else:
+        if pcodega and ccode:
+            barcode = pcodega + 'XX' + ccode
+            temp_test = barcode == values['barcode']
+        else:
+            temp_test = False
+    test = test and temp_test
+    if not temp_test:
+        error = error + 'barcode,\n'
+
+    # third: languages
+    temp_test = True
+    for lang in line_vals['test_letter_languages_id'].split(','):
+        temp_test = temp_test and lang in values['lang']
+    line_nber_lang = len(line_vals['test_letter_languages_id'].split(','))
+    csv_nber_lang = len(values['lang'].split(','))
+    temp_test = temp_test and line_nber_lang == csv_nber_lang
+
+    test = temp_test and test
+    if not temp_test:
+        error = error + 'lang,\n'
+
+    return test, error[:-2]
```

Comment:
Add a comment explaining you remove end of line in error string. You could avoid to return `test` variable and instead check outside of the func:

```
if error:
    .... (means an error was detected and error string is not empty)
else:
    .... (no error detected)
```